### ISSUE TITLE
VS2017: Undesired reformatting when uncommenting a section

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#29647]
I'm working in C# and I have a line of code like this:

	foo( a, b );

I decide to comment it, using the toolbar button or CTRL-K, CTRL-C. It becomes:

	//foo( a, b );

Then I want to revert the operation, using the opposite function. VisualStudio decides that removing the comment marker is not enough, and the line becomes:

	foo(a, b);

Please do NOT reformat my code without my consent, ever!

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/324827/vs2017-undesired-reformatting-when-uncommenting-a.html
VSTS ticketId: 674251_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_

Jinu Joseph [MSFT] solved on 9/3/2018, 04:16 AM (4 hours ago), 0 votes: <p>Thank you for your feedback! This issue appears to be identical to  <a target='_blank' href="https://github.com/dotnet/roslyn/issues/29482">https://github.com/dotnet/roslyn/issues/29482</a>  We recommend that you vote on and follow the earlier reported issue for updates and fix notifications.</p>

**Comment by sharwell** [comment]
I am taking two proposals to a design meeting (along with any others that might get proposed):

1. Allow the **Tools &rarr; Options.. &rarr; Text Editor &rarr; C# &rarr; Code Style &rarr; Formatting &rarr; General** page to include a new option "Automatically format on uncomment"
2. Update the behavior of the *existing* option "Automatically format on paste" to cover the uncomment case as well

Once we figure out a design, we can mark this up-for-grabs, though it's a rare enough request that the implementation itself will almost certainly need to be provided via a community pull request before it moves forward.

**Comment by wiz78** [comment]
I’ve seen lots of people asking for VS not to reformat code without permission (such as when it used to reformat razor code on paste even if you switched every possible reformatting options off)

I can’t understand why you think that reformatting someone else’s code is ok, whatever the circumstances are, unless auto-formatting has been explicitly enabled. Unless there’s a way to turn it off, auto formatting should **never** occur, IMHO.

Also, in this very specific case, it’s a regression: older VS versions didn’t reformat code when commenting/uncommenting lines.

**Comment by wiz78** [comment]
...an example to rule them all: 
https://github.com/dotnet/roslyn/issues/8263

**Comment by CyrusNajmabadi** [comment]
> I can’t understand why you think that reformatting someone else’s code is ok, whatever the circumstances are, unless auto-formatting has been explicitly enabled. 

I disagree, and taking that position would result in very large regressions in behavior that people have become enormously accustomed to.  I'm fine with formatting having opt-out knobs.  But requiring that 'auto formatting [be] explicitly enabled' would be lead to a totally regressed experience in many scenarios for customers who have grown used to this since VS2002.

> Also, in this very specific case, it’s a regression: older VS versions didn’t reformat code when commenting/uncommenting lines.

Sure.  But it's also worth nothing that this was done intentionally as many people complained when uncommenting that commented code from elsewhere was not appropriately formatted.  This often happened when code was commented out, then code around those comments was changed (for example, introducing multiple if-statements, or other sorts of surround-with constructs.  Then, when the code was uncommented, it would now be at a vastly wrong indentation level.  

Now, i'm amenable to the position that this be something you can disable.  I think opt-out is perfectly fine.  However, there are very strong reasons why formatting happens, and why it's a good default for many circumstances.

**Comment by wiz78** [comment]
yes, perhaps "explicitly" was the wrong word: "controlled by a knob" is what I meant. The default value is not so important to me, the ability to make a choice is.

**Comment by CyrusNajmabadi** [comment]
That works for me as well :)

**Comment by dibarbet** [comment]
@sharwell this one as well, would you happen to recall what the design review outcome of this one was as well?

**Comment by sharwell** [comment]
Design review conclusion: we would accept a pull request to implement a new option on this page, which indicates whether formatting should be applied when code is uncommented:

![image](https://user-images.githubusercontent.com/1408396/53419261-241d6400-399f-11e9-8efb-93c91564d75b.png)
